core/components/addressable_pq.hpp

```diff
@@ -0,0 +1,199 @@
+ */
+template <typename KeyType, typename ValueType>
+struct addressable_priority_queue {
+    explicit addressable_priority_queue(int deg_log2) : degree{1 << deg_log2} {}
+
+    /**
+     * Inserts the given key-value pair into the PQ.
+     * Duplicate keys are allowed, they may be returned in an arbitrary order.
+     * @returns a handle for the pair to be used when modifying the key.
+     */
+    std::size_t insert(KeyType key, ValueType value)
+    {
+        m_keys.push_back(key);
+        m_values.push_back(value);
+        auto handle = next_handle();
+        m_handles.push_back(handle);
+        if (handle == m_handle_pos.size())
+            m_handle_pos.push_back(size() - 1);
+        else
+            m_handle_pos[handle] = size() - 1;
+        sift_up(size() - 1);
+        return handle;
+    }
+
+    /**
+     * Updates the key of the pair with the given handle.
+     */
+    void update_key(std::size_t handle, KeyType new_key)
+    {
+        auto pos = m_handle_pos[handle];
+        GKO_ASSERT(pos < size());
+        GKO_ASSERT(m_handles[pos] == handle);
+        auto old_key = m_keys[pos];
+        m_keys[pos] = new_key;
+        if (old_key < new_key) {
+            sift_down(pos);
+        } else {
+            sift_up(pos);
+        }
+    }
+
+    /**
+     * Returns the minimum key from the queue.
+     */
+    KeyType min_key() const { return m_keys[0]; }
+
+    /**
+     * Returns the value belonging to the minimum key from the queue.
+     */
+    ValueType min_val() const { return m_values[0]; }
+
+    /**
+     * Returns the key-value pair with the minimum key from the queue.
+     */
+    std::pair<KeyType, ValueType> min() const { return {min_key(), min_val()}; }
+
+    /**
+     * Removes the key-value pair with the minimum key from the queue.
+     */
+    void pop_min()
+    {
+        swap(0, size() - 1);
+        m_keys.pop_back();
+        m_values.pop_back();
+        auto old_handle = m_handles.back();
+        m_handles.pop_back();
+        sift_down(0);
+    }
+
+    /**
+     * Returns the number of key-value pairs in the queue.
+     */
+    std::size_t size() const { return m_keys.size(); }
+
+    /**
+     * Returns true if and only if the queue has size 0.
+     */
+    bool empty() const { return size() == 0; }
+
+    void reset()
+    {
+        m_keys.clear();
+        m_values.clear();
+        m_handles.clear();
+        m_handle_pos.clear();
+    }
+
+private:
+    // constexpr static int degree = 1 << Degree_Log2;
+
+    std::size_t parent(std::size_t i) const { return (i - 1) / degree; }
+
+    std::size_t first_child(std::size_t i) const { return degree * i + 1; }
+
+    void swap(std::size_t i, std::size_t j)
+    {
+        std::swap(m_keys[i], m_keys[j]);
+        std::swap(m_values[i], m_values[j]);
+        std::swap(m_handles[i], m_handles[j]);
+        std::swap(m_handle_pos[m_handles[i]], m_handle_pos[m_handles[j]]);
+    }
+
+    void sift_down(std::size_t i)
```

Comment:
Maybe add some documentation to this and to sift_up. Even if it is a private method, it would improve understandability of the code.